File: python/src/biosimulant_model_compatibility_standard/units.py
"""Unit expressions: parse a UCUM code, and say whether two units convert.

The heavy lifting happened at build time: scripts/build_ucum_table.py resolved every UCUM unit to
a factor, a dimension vector and, where the unit is affine, an offset. This module parses an
expression over that table and answers one question: can a value in unit A be expressed in unit B,
and by what factor?

Three answers are distinct, and the standard needs all three:

- convertible, with a factor and possibly an offset;
- not convertible, because the dimensions differ (a known contradiction);
- not decidable, because a unit is unparseable, arbitrary or logarithmic (UNKNOWN, never a pass).

See decision D1.
"""
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Any

TOKEN = re.compile(r"""
    (?P<open>\()
  | (?P<close>\))
  | (?P<operator>[./])
  | (?P<annotation>\{[^}]*\})
  | (?P<number>\d+)(?=$|[./(){}]|\{)
  | (?P<code>\[[^\]]*\]|[^./(){}\s]+?)(?P<exponent>[+-]?\d+)?(?=$|[./(){}]|\{)
""", re.VERBOSE)


class UnitError(ValueError):
    """The expression cannot be interpreted, so no comparison may be inferred from it."""


@dataclass(frozen=True)
class Quantity:
    """A parsed unit: a factor over the base units, plus its dimension."""

    factor: float
    dimension: tuple[int, ...]
    offset: float = 0.0
    arbitrary_codes: frozenset[str] = frozenset()
    codes: frozenset[str] = frozenset()

    @property
    def dimensionless(self) -> bool:
        return not any(self.dimension)


@dataclass(frozen=True)
class Conversion:
    """How to get from one unit to another. value_in_target = value * factor + offset."""

    factor: float
    offset: float
    affine: bool
    lossless: bool = True


def parse_unit(expression: str, table: dict[str, Any]) -> Quantity:
    """Parse a UCUM expression against a flattened table. Raises UnitError when it cannot."""
    if not isinstance(expression, str) or not expression.strip():
        raise UnitError("empty unit")
    text = table.get("aliases", {}).get(expression.strip(), expression.strip())
    units, prefixes = table["units"], table["prefixes"]
    width = len(table["dimensions"])
    position = 0
    arbitrary: set[str] = set()
    used: set[str] = set()
    offsets: list[float] = []
    components = 0
    operators = 0

    def resolve(code: str) -> tuple[float, list[int]]:
        entry = units.get(code)
        if entry is None:
            for length in (2, 1):
                prefix, rest = code[:length], code[length:]
                candidate = units.get(rest)
                if prefix in prefixes and candidate is not None and candidate.get("metric"):
                    entry = {**candidate, "factor": prefixes[prefix] * candidate["factor"]}
                    break
        if entry is None:
            raise UnitError(f"unknown unit code {code!r}")
        used.add(code)
        if entry.get("arbitrary"):
            arbitrary.add(code)
        if entry.get("offset"):
            offsets.append(entry["offset"])
        return entry["factor"], entry["dim"]

    def component() -> tuple[float, list[int]]:
        nonlocal position, components
        match = TOKEN.match(text, position)
        if not match:
            raise UnitError(f"cannot parse {expression!r}")
        position = match.end()
        if match.group("open"):
            factor, dimension = sequence()
            closing = TOKEN.match(text, position)
            if not closing or not closing.group("close"):
                raise UnitError(f"unbalanced parentheses in {expression!r}")
            position = closing.end()
            return factor, dimension
        if match.group("annotation"):
            return 1.0, [0] * width
        if match.group("operator") or match.group("close"):
            raise UnitError(f"unexpected operator in {expression!r}")
        if match.group("number"):
            # A bare integer is a factor in its own right: 86400 is not 8 to the 6400th.
            return float(match.group("number")), [0] * width
        code, exponent = match.group("code"), int(match.group("exponent") or 1)
        components += 1
        factor, dimension = resolve(code)
        trailing = TOKEN.match(text, position)
        if trailing and trailing.group("annotation"):
            # An annotation binds to the unit before it and carries no semantics: g{DW} is grams.
            position = trailing.end()
        return factor ** exponent, [value * exponent for value in dimension]

    def sequence() -> tuple[float, list[int]]:
        nonlocal position, operators
        leading = TOKEN.match(text, position)
        if leading and leading.group("operator") == "/":
            position = leading.end()
            operators += 1
            right_factor, right_dimension = component()
            factor, dimension = 1.0 / right_factor, [-value for value in right_dimension]
        else:
            factor, dimension = component()
        while position < len(text):
            match = TOKEN.match(text, position)
            if not match or not match.group("operator"):
                break
            position = match.end()
            operators += 1
            operator = match.group("operator")
            right_factor, right_dimension = component()
            if operator == ".":
                factor *= right_factor
                dimension = [a + b for a, b in zip(dimension, right_dimension)]
            else:
                factor /= right_factor
                dimension = [a - b for a, b in zip(dimension, right_dimension)]
        return factor, dimension

    factor, dimension = sequence()
    if position != len(text):
        raise UnitError(f"trailing input in {expression!r}")
    if offsets and (len(offsets) > 1 or operators or components != 1):
        # Celsius converts on its own. A degree Celsius per minute does not: the offset has no
        # meaning once the unit is combined with anything else.
        raise UnitError(f"{expression!r} combines an affine unit with other units")
    return Quantity(factor, tuple(dimension), offsets[0] if offsets else 0.0, frozenset(arbitrary), frozenset(used))


def convert_unit(source: str, target: str, table: dict[str, Any]) -> Conversion | None:
    """Return how to convert source into target, or None when the dimensions differ.

    Raises UnitError when the question cannot be decided: an unparseable code, or an arbitrary
    unit such as [IU] or [PFU], which UCUM defines as commensurable with nothing but itself.
    """
    if source == target:
        return Conversion(1.0, 0.0, affine=False)
    left, right = parse_unit(source, table), parse_unit(target, table)
    if left.arbitrary_codes or right.arbitrary_codes:
        if left.arbitrary_codes != right.arbitrary_codes:
            raise UnitError("arbitrary units are commensurable only with themselves")
    if left.dimension != right.dimension:
        return None
    factor = left.factor / right.factor
    offset = (left.offset - right.offset) / right.factor
    return Conversion(factor, offset, affine=offset != 0.0)

File: typescript/src/units.ts
// Unit expressions: parse a UCUM code, and say whether two units convert.
//
// The resolution of UCUM's definition chains happened at build time
// (scripts/build_ucum_table.py). This module parses an expression over that flat table and
// answers one question: can a value in unit A be expressed in unit B, and by what factor?
//
// Three answers are distinct, and the standard needs all three: convertible with a factor and
// possibly an offset; not convertible because the dimensions differ; and not decidable, because a
// unit is unparseable, arbitrary or logarithmic. The last must never resolve as a pass.
//
// Mirrors python/src/biosimulant_model_compatibility_standard/units.py.

export interface UnitTable {
  dimensions: string[];
  prefixes: Record<string, number>;
  aliases?: Record<string, string>;
  units: Record<string, { factor: number; dim: number[]; metric?: boolean; offset?: number; arbitrary?: boolean }>;
}

export interface Quantity {
  factor: number;
  dimension: number[];
  offset: number;
  arbitraryCodes: string[];
  codes: string[];
}

export interface Conversion {
  factor: number;
  offset: number;
  affine: boolean;
  lossless: boolean;
}

export class UnitError extends Error {}

const TOKEN = /(\()|(\))|([./])|(\{[^}]*\})|(\d+)(?=$|[./(){}]|\{)|(\[[^\]]*\]|[^./(){}\s]+?)([+-]?\d+)?(?=$|[./(){}]|\{)/y;

export function parseUnit(expression: string, table: UnitTable): Quantity {
  if (typeof expression !== "string" || !expression.trim()) throw new UnitError("empty unit");
  const text = table.aliases?.[expression.trim()] ?? expression.trim();
  const width = table.dimensions.length;
  const arbitrary = new Set<string>();
  const used = new Set<string>();
  const offsets: number[] = [];
  let position = 0;
  let components = 0;
  let operators = 0;

  const zero = () => new Array(width).fill(0);

  const next = () => {
    TOKEN.lastIndex = position;
    const match = TOKEN.exec(text);
    if (!match) throw new UnitError(`cannot parse ${expression}`);
    position = TOKEN.lastIndex;
    return match;
  };

  const peek = () => {
    TOKEN.lastIndex = position;
    return TOKEN.exec(text);
  };

  const resolve = (code: string): [number, number[]] => {
    let entry = table.units[code];
    if (!entry) {
      for (const length of [2, 1]) {
        const prefix = code.slice(0, length), rest = code.slice(length);
        const candidate = table.units[rest];
        if (table.prefixes[prefix] !== undefined && candidate?.metric) {
          entry = { ...candidate, factor: table.prefixes[prefix] * candidate.factor };
          break;
        }
      }
    }
    if (!entry) throw new UnitError(`unknown unit code ${code}`);
    used.add(code);
    if (entry.arbitrary) arbitrary.add(code);
    if (entry.offset) offsets.push(entry.offset);
    return [entry.factor, entry.dim];
  };

  const component = (): [number, number[]] => {
    const match = next();
    if (match[1]) {
      const inner = sequence();
      const closing = next();
      if (!closing[2]) throw new UnitError(`unbalanced parentheses in ${expression}`);
      return inner;
    }
    if (match[4]) return [1, zero()];
    if (match[3] || match[2]) throw new UnitError(`unexpected operator in ${expression}`);
    // A bare integer is a factor in its own right: 86400 is not 8 to the 6400th.
    if (match[5]) return [Number(match[5]), zero()];
    const code = match[6], exponent = match[7] ? Number.parseInt(match[7], 10) : 1;
    components += 1;
    const [factor, dimension] = resolve(code);
    const trailing = peek();
    // An annotation binds to the unit before it and carries no semantics: g{DW} is grams.
    if (trailing && trailing[4]) position = TOKEN.lastIndex;
    return [factor ** exponent, dimension.map((value) => value * exponent)];
  };

  const sequence = (): [number, number[]] => {
    let factor: number, dimension: number[];
    const leading = peek();
    if (leading && leading[3] === "/") {
      position = TOKEN.lastIndex;
      operators += 1;
      const [rightFactor, rightDimension] = component();
      factor = 1 / rightFactor;
      dimension = rightDimension.map((value) => -value);
    } else {
      [factor, dimension] = component();
    }
    while (position < text.length) {
      const match = peek();
      if (!match || !match[3]) break;
      position = TOKEN.lastIndex;
      operators += 1;
      const operator = match[3];
      const [rightFactor, rightDimension] = component();
      if (operator === ".") {
        factor *= rightFactor;
        dimension = dimension.map((value, index) => value + rightDimension[index]);
      } else {
        factor /= rightFactor;
        dimension = dimension.map((value, index) => value - rightDimension[index]);
      }
    }
    return [factor, dimension];
  };

  const [factor, dimension] = sequence();
  if (position !== text.length) throw new UnitError(`trailing input in ${expression}`);
  if (offsets.length && (offsets.length > 1 || operators || components !== 1)) {
    // Celsius converts on its own. A degree Celsius per minute does not: the offset has no
    // meaning once the unit is combined with anything else.
    throw new UnitError(`${expression} combines an affine unit with other units`);
  }
  return { factor, dimension, offset: offsets[0] ?? 0, arbitraryCodes: [...arbitrary].sort(), codes: [...used].sort() };
}

/**
 * How to convert source into target, or null when the dimensions differ.
 * Throws UnitError when the question cannot be decided.
 */
export function convertUnit(source: string, target: string, table: UnitTable): Conversion | null {
  if (source === target) return { factor: 1, offset: 0, affine: false, lossless: true };
  const left = parseUnit(source, table), right = parseUnit(target, table);
  if (left.arbitraryCodes.length || right.arbitraryCodes.length) {
    if (JSON.stringify(left.arbitraryCodes) !== JSON.stringify(right.arbitraryCodes)) {
      throw new UnitError("arbitrary units are commensurable only with themselves");
    }
  }
  if (JSON.stringify(left.dimension) !== JSON.stringify(right.dimension)) return null;
  const factor = left.factor / right.factor;
  const offset = (left.offset - right.offset) / right.factor;
  return { factor, offset, affine: offset !== 0, lossless: true };
}
